src/nodes.coffee

```diff
@@ -1152,25 +1192,75 @@ exports.Class = class Class extends Base
     args  = []
     o.classScope = func.makeScope o.scope
 
+    superClass = null
+    if @parent
+      superClass = o.classScope.freeVariable 'superClass', reserve: no
+      func.params.push new Param new IdentifierLiteral superClass
+      args.push @parent
+
     @hoistDirectivePrologue()
     @setContext name
     @walkBody name, o
-    @ensureConstructor name
+    @ensureConstructor()
     @addBoundFunctions o
+    @ensureConstructorSuperCall() if @parent
+
+    esClass = new EsClass name, superClass, @ctor
+    @hoistConstructorComment esClass
+
     @body.spaced = yes
+    @body.expressions.unshift @directives..., esClass.placeholder
     @body.expressions.push lname
+    @body.expressions.splice @body.expressions.indexOf(@ctor), 1, esClass
 
-    if @parent
-      superClass = new IdentifierLiteral o.classScope.freeVariable 'superClass', reserve: no
-      @body.expressions.unshift new Extends lname, superClass
-      func.params.push new Param superClass
-      args.push @parent
+    klass  = new Parens new Call func, args
+    klass  = new Assign @variable, klass, null, { @moduleDeclaration } if @variable
+    result = klass.compileToFragments o
+    esClass.placeholder.updateFragments result
+    result
+
+#### ES2015+ Classes
+
+# An EsClass represents an ECMAScript class definition.
+# Initialize an **EsClass** with its name, an optional superclass,
+# and an optional constructor.
+exports.EsClass = class EsClass extends Base
+  class Placeholder extends Base
+    constructor: ->
+      @nonce     = {}
+      @fragments = []
+
+    isStatement: YES
+
+    compileToFragments: (o) ->
+      @nonce
+
+    updateFragments: (fragments) ->
+      fragments.splice fragments.indexOf(@nonce), 1, @fragments...
+
+  children: ['ctor', 'comment']
 
-    @body.expressions.unshift @directives...
+  constructor: (@name, @parent, @ctor, @comment) ->
+    @placeholder = new Placeholder
 
-    klass = new Parens new Call func, args
-    klass = new Assign @variable, klass, null, { @moduleDeclaration } if @variable
-    klass.compileToFragments o
+  isStatement: YES
+
+  compileNode: (o) ->
+    code = @placeholder.fragments
+
+    code.push @makeCode "#{@tab}class #{@name} "
+    code.push @makeCode "extends #{@parent} " if @parent
+    code.push @makeCode '{'
+
+    unless @ctor.params.length is 0 and @ctor.body.isEmpty() and not @comment
+      o.indent += TAB
+      code.push @comment.compileToFragments(o)... if @comment
+      code.push @makeCode('\n'), @ctor.compileToFragments(o)..., @makeCode "\n#{@tab}"
+
+    code.push @makeCode '}'
+    code.push @makeCode "\n#{@tab}#{@name}.__super__ = #{@parent}.prototype;" if @parent
+
+    []
```

Comment:
At the very least, please add a comment near the returned empty array that explains this.

Look at my commit where I removed extra empty lines. Could you put some code there that helps with this?